### DISCUSSION TITLE
Syllabus updates for FullStack

### DISCUSSION THREAD

**Discussion by alesanchezr** [breatheco-de/breatheco-de#6952] [Syllabus]
No lesson on HIGHER ORDER FUNCTIONS?

WHAT IS JAVASCRIPT LESSON

- [ ] Break up Intro to JS lesson - too big! Intro to JS should be a short intro lesson introducing JS/runtime env and how it works. No code. Followed by: 
- [ ] Individual lessons variables
- [x] [Individual lesson on conditionals](https://4geeks.com/lesson/conditionals-in-programing-coding)
- [x] [Individual lesson on arrays](https://4geeks.com/lesson/what-is-an-array-define-array)
- [x] [Individual lesson on functions](https://4geeks.com/lesson/working-with-functions)
- [x] The Replit example under Global Variables is broken (Remove? Replace?) 

JS - no lesson on strings and number manipulation/methods; 
- [x] Here is a lesson  about [strings in javascript](https://4geeks.com/lesson/strings-in-javascript)
- [ ] We are missing the lesson about number manipulation but I'm not sure how much can we say about this.

CONDITIONALS LESSON
More/better curriculum on ternary expressions and logical operators, esp. how AND and OR behave to evaluate course of action.
- [X] here is a [lesson about conditionals](https://4geeks.com/lesson/conditionals-in-programing-coding). 

JAVASCRIPT BEGINNER TUTORIAL 
- [ ] Break up exercise set Javascript Beginner Tutorial into topics: e.g. Strings, Numbers and other Data Types in one set; Conditionals in another set, etc. TO MATCH LESSONS ORGANIZATION.

The JS beginner learnpack tutorial contains 24 exercises right now, I worry that adding more could make it too long for the prework.
Also, we have them broken in Functions, Loops, OOP. 
Maybe we could one learnpack about "strings and conditions" and have several exercises about it?

-  [ ] Agreed: Number 24 in that set is about Objects - it should be taken out and placed after Objects is covered as a topic. 
- [ ] Agreed: Exercises in this set should only use function declarations, no arrow functions as they have not been covered at this stage. 

WORKING WITH FUNCTIONS LESSON

- [ ] Question: Last example includes arrays and higher order functions but neither has been introduced to them yet. 
You mean to remove exercise 9?

- [ ] Agreed: No curriculum on ARROW FUNCTIONS! We need to include some - possibly in Functions lesson. 

JAVASCRIPT FUNCTIONS TUTORIAL 
Include more exercises on arrow functions - make the students practice the different shorthand versions of an arrow function. Make them use arrow function as a callback function. 
Make sure arrays are not mentioned since they are not introduced - replace ex 9. 

EXCUSE GENERATOR project uses arrays and loops BEFORE they are taught. If it should stay as the project for functions, then it should focus more on practicing functions and other concepts taught up to that point.

DOMAIN NAME GENERATOR is very similar - only difference is that Domain generator requires use of nested loops 

OOP
Object Oriented Programming not included? We need to have a major lesson on that - objected oriented code, object literals, accessing deeply nested objects. 

WHAT IS AN ARRAY LESSON
Replace For…in example with For…of - For in is not recommended for Arrays, rather for Objects 
Include example of using .entries() method in For…of
Replit examples do not load on the page - the links work though

CONDITIONAL RENDERING PROJECT
Should come AFTER DOM and Events are taught - involves way too much of those concepts 

LEARN ARRAYS AND LOOPS EXERCISES

We need curriculum on DESTRUCTURING and SPREAD OPERATOR.

**Comment by arnaldoperez** [comment]
## Building Starwars Data ~~Model~~ Base

- Agregar lección de modelado de datos
Reseñar la importancia del modelado de datos, buenas prácticas, los problemas que ocasionan modelos de datos sin normalizar. Mostrar como aplicar las formas normales N1, N2 y N3 con un caso de estudio.
- Extender el proyecto de Star Wars model
Agregar el desarrollo de funcionalidades CRUD en un conjunto de funciones preexistentes vacías, que incluya además consultas entre distintas tablas. Se pueden realizar pruebas unitarias para validar las funciones.
-  Ubicar después de “Intro to Python”

## Family API

- Building RESTful API's using Flask
- Todo List API with Python Flask Interactive
- Family Static API with Flask

## Star Wars blog - Building endpoints

- Solo hacer el proyecto “Build a StarWars REST API”
- Ubicar después de la clase “Family API”

## User Stories

- Cambiar el entregable de la clase  por titulo de proyecto, wireframe y lista de historias de usuarios
- 📗 Requerimientos del proyecto final (Knowledge database)
- 🧪Historias de usuario del proyecto

## (Módulo nuevo)Proyecto final
- 🧪Arranque del proyecto: Crear el repositorio, invitar al equipo, se crea el kanban (Trello, GH Project, etc).
- 🧪Modelado de datos del proyecto
- 📗 Repositorio de apis externas para realizar proyectos
- 📗 Material de apoyo

**Comment by devAbreu** [comment]
## (Módulo nuevo) Debugging

- 🧪 Introducción al debugging
- 🧪 Herramientas de debugging
- 🧪 Técnicas de debugging en el frontend
- 🧪 Técnicas de debugging en el backend
- 🧪 Estrategias avanzadas de debugging
- 📗 Material de apoyo

**Comment by arnaldoperez** [comment]
## Módulo 30 - Historias de usuario

- 📗 [¿Por qué construir un proyecto final?](https://4geeks.com/es/lesson/porque-construir-un-proyecto-final)
- 📗 [Creando Historias de usuarios](https://4geeks.com/lesson/user-stories-examples)
- 📗 [Introducción al desarrollo profesional y ágil](https://4geeks.com/lesson/agile-development)
- [x] 📗 [¿Cómo debe ser un proyecto Full-Stack para tu portafolio?]
- [x] 📗 [¿Cómo debe ser un proyecto de DataScience y Machine Learning para tu portafolio?]
- [x]  🧪 Historias de usuario (Product Backlog) y alcance del proyecto final

## Módulo 31 - Diseño y Wireframes
- [x] 📗 Introducción al diseño UI/UX
- [x]  🧪 Wireframes del proyecto final

# Módulo 32: Arranque del proyecto final

- [x] 📗 Repositorio de apis externas para realizar proyectos
- [x] 📗 Librerías populares
- [x] 📗 Modelado de base de datos
- [x] 🧪Modelado de datos de un Zoológico https://app.quickdatabasediagrams.com/#/
- [x] 🧪Modelado de datos del proyecto
